### PR TITLE
theano support for 5d tensor in placeholder and rnn

### PR DESCRIPTION
**Pull request by jeammimi**
Added the possibility to create 5D tensor in theano,
correction of a bug in rnn which prevented output shape to
be different from input shape. (for 5D tensors)

### CODE REVIEW

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -51,6 +51,9 @@ def placeholder(shape=None, ndim=None, dtype=_FLOATX, name=None):
         return T.tensor3(name=name, dtype=dtype)
     elif ndim == 4:
         return T.tensor4(name=name, dtype=dtype)
+    elif ndim == 5:
+        dtensor5 = T.TensorType(dtype, (False,)*5)
+        return dtensor5(name)
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Then you should apply the same modifications in the TensorFlow backend, for consistency.

**Comment by jeammimi** [comment]
Tensorflow was not limited to 4d tensors for the placeholder

I modified rnn for tensorflow to make it works with 5d tensor.
I didn't tested the mask option for rnn for more than 3d tensor  in
both backend.

**Comment by fchollet** [comment]
Regarding the TensorFlow changes: I'm worried about performance regressions, since you are doing heavier operations. Can you comment about that? For instance, a 15% performance regression had been previously introduced while fixing masking (which I just solved).

File: keras/backend/theano_backend.py (hunk 1 above)
Comment (nouiz) [inline keras/backend/theano_backend.py]: Here is the generic version that will work for any ndim:
return T.TensorType(dtype, (False,)*ndim)(name)

So this whole if/elif/.../else will disapear.

Comment (fchollet) [inline keras/backend/theano_backend.py]: @nouiz btw, is it possible to instantiate a placeholder tensor without a specific ndim, like it is in TensorFlow (equivalent to `tf.placeholder()`)?

Comment (nouiz) [inline keras/backend/theano_backend.py]: Yes, the code I give do that.

`T.TensorType(dtype, (False,)*ndim)` create a Type instead that describe the type. Then with it, you can instantiate a Theano variable by calling it `()`. You can give a variable name as arguments `(name)`

Comment (fchollet) [inline keras/backend/theano_backend.py]: Excellent, thanks.

Comment (fchollet) [inline keras/backend/theano_backend.py]: But about the `broadcastable` bool tuple: is it required to build the graph? Could we have a tensor that can accept data of *variable* ndim (like `tf.placeholder()`, which is ndim-agnostic)? I'm asking because occasionally the ndim of certain inputs isn't known at compilation time, which is difficult to work around.

Comment (nouiz) [inline keras/backend/theano_backend.py]: Currently, Theano mostly make this mandatory to have the number of dimensions known when building the graph. You could put a Generic() variable, but you can't do much stuff with this. You could clone the graph and replace it later, but it is getting complicated.

When do you have this problem of not knowing the number of dimensions when you build the graph? I don't recall having saw this problem. You don't need to know the shape, just the number of dimensions and if you want to broadcast on a giving dimensions or not.

Comment (fchollet) [inline keras/backend/theano_backend.py]: The latest issue I've seen was knowing whether sample weights would be 2D (sample-wise) or 3D (timestep-wise). Here: https://github.com/fchollet/keras/blob/master/keras/models.py#L470

The current workaround consist in making it possible to specify "temporal" or None as `sample_weight_mode`, which adds quite a bit of unnecessary complexity to the API. Another possible workaround would have been to compile the graph only when `fit` is called (and therefore when we have access to the`sample_weight` data passed by the user) but that would be somewhat hacky (would rely on the user passing consistent weights everytime) and it would require changing the Keras architecture to a large extent.

**Comment by jeammimi** [comment]
I think the code can be improved:
if  mask is None, I think we can avoid doing any switch operation.
We can keep the loop by creating a mask = [True] * timestep, and 
then remove all the switch operation by testing with a if .

if mask is not None, I can get out of the loop the exand_dims and tile operation,
but it will probably require a little bit of extra memory.

Right know for me the operation tf.cast(mask, tf.bool) does not work so I can not really check
the performance.

I will do a new version af rnn that implement these changes today.

**Comment by jeammimi** [comment]
So I separated the masked and regular loop. like that we are sure that
the mask operation does not slow down the regular inputs.

For the masked operation, we have to tile the mask on all tha dimensions,
so the only extra operation introduce with respect  to the original implementation
that was working on three dimensions is a squeze  which remove all the dimension equal 
to one. I don't think it consume some time. I can remove this if I knew how to make a copy
of a tensorflow tensor.

**Comment by fchollet** [comment]
This PR needs to be rebased to master. But more importantly, can you remind me why it's important to allow for a mask with a different shape from the input? Masks are generated either by Embedding (3D, same as input) or by a Masking layer (nD, same as input).

**Comment by jeammimi** [comment]
This pull request can be clause because fchollet, made same kind of modification
